pkg/steps/source.go

```diff
@@ -368,6 +370,52 @@ func isBuildPhaseTerminated(phase buildapi.BuildPhase) bool {
 	return true
 }
 
+func handleBuilds(ctx context.Context, buildClient BuildClient, build buildapi.Build) error {
+	var wg sync.WaitGroup
+	errChan := make(chan error)
+
+	for _, build := range constructMultiArchBuilds(build, buildClient.NodeArchitectures()) {
```

Comment:
What about handling context cancellation like the following?
```golang
for _, build := range constructMultiArchBuilds(build, buildClient.NodeArchitectures()) {
	select {
		case <-ctx.Done():
		default:
	}
	wg.Add(1)
	go func(b buildapi.Build) {
		defer wg.Done()

		if err := handleBuild(ctx, buildClient, b); err != nil {
			errChan <- fmt.Errorf("error occurred handling build %s: %w", b.Name, err)
		}

	}(build)
}
```
This would prevent some extra calls of `handleBuild` if  `ctx` was cancelled before the loop started or even in the middle of it.

Reply:
This should be already handled in `handleBuild`. Also, once the context is canceled we need to cancel all builds and not continue.

Reply:
Yes the context is also handled downstream and, once canceled, every build cancels as well in both the implementation, but as an example:
```golang
for _, build := []string{"amd64", "x86", "arm64"} {
	go handleBuild(ctx, build)
}
```
it keeps processing `"x86"` and `"arm64"` even if the context get canceled after the first iteration `"amd64"`.
On the contrary:
```golang
for _, build := []string{"amd64", "x86", "arm64"} {
	select {
		case <-ctx.Done():
		    break
                default:
	}
	go handleBuild(ctx, build)
}
```
it stops right away after `"amd64"`.

Reply:
That case will never happen either way. We need to spawn all of the builds simultaneously and if the context is canceled then each of the goroutines should handle the cancelation separately.

Reply:
It can definitely happen, but it's unlikely that cancellation will happen between the creation of those threads.  In this particular case, it's fine to let each `handleBuild` return an error as soon as the cancellation happens.

But there is a deeper problem, of which this is not the first instance, but may be the worst.  `handleBuild` ultimately uses the `client-go` infrastructure, and each of these calls being completely independent means each will have its own informer, index, watch, etc.  This is already the case when we have multiple unrelated builds across steps, but here it is even more outrageous since they are are being created in a single place.